### PR TITLE
Add subdomain name duplication check

### LINKED ISSUE #23735: Disable two (duplicate / overlapping) subdomain IDs / names for a given subdomain name / ID by default
## Reason
One of these combinations is allowed and is not intuitive and giving trouble to modelers downstream

## Design
Add a parameter and disallow it by default

## Impact
Easier use of Moose meshes

### PR DESCRIPTION
**Pull request by MengnanLi91**
close #23735

## Reason

The subdomains with different IDs are allowed to have same subdomain name in current version. This is not intuitive and seems giving trouble to modelers downstream.

## Design

A subdomain name duplication check function is added to `MooseMesh`. It will throw out an error message if the name duplication is found for the subdomains with different IDs.

## Impact

Fix the issue #23735

### CODE REVIEW

**Comment by GiudGiud** [comment]
Let's see who was using that and if we need to allow it

File: framework/src/mesh/MooseMesh.C [hunk 1]
```diff
@@ -3699,3 +3702,22 @@ MooseMesh::lengthUnit() const
...
+  for (const auto & sbd_id : _mesh_subdomains)
+  {
+    std::string sub_name = getSubdomainName(sbd_id);
+    if (!sub_name.empty() && subdomain.count(sub_name) > 0)
+      mooseError("The subdomain name ",
+                 sub_name,
+                 " has already been used for the subdomain with ID=",
+                 subdomain[sub_name],
+                 ", Please rename the subdomain with ID=",
+                 sbd_id);
```
Comment (GiudGiud) [inline framework/src/mesh/MooseMesh.C:3719]: let s reword this to be "is used by both subdomains with ID". ... " and ID " ...
Please rename one of them).

we dont know which one is the best to rename

**Review by GiudGiud — COMMENTED** [review]
@socratesgorilla if you could please make a patch to that SAM input that uses two blocks with the same subdomain name and different IDs

**Comment by travismui** [comment]
@GiudGiud @socratesgorilla FYI I took care of patching those tests in SAM, ref [SAM/SAM!717](https://git-nse.egs.anl.gov/SAM/SAM/-/merge_requests/717)

**Comment by GiudGiud** [comment]
Awesome! Thanks a lot

**Review by GiudGiud — APPROVED** [review]
Thanks for the new check! We'll invalidate on the next SAM master merge and get this in

@aprilnovak fyi on name vs id duplication
